core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -410,115 +410,77 @@ private[fs2] object Algebra {
+              .Eval[Algebra[G, X, ?], Q](alg)
+              .transformWith(res =>
+                translateStep(view.next(res.asInstanceOf[Result[y]]), isMainLevel))
 
-        }
-
-    }
-
-  private def translate0[F[_], G[_], O](
-      fK: F ~> G,
-      s: FreeC[Algebra[F, O, ?], Unit],
-      concurrent: Option[Concurrent[G]]
-  ): FreeC[Algebra[G, O, ?], Unit] =
-    s.viewL match {
-      case _: FreeC.Result.Pure[Algebra[F, O, ?], Unit] =>
-        FreeC.pure[Algebra[G, O, ?], Unit](())
+          view.step match {
+            case output: Algebra.Output[F, X] =>
+              Algebra.output[G, X](output.values).transformWith {
+                case r @ Result.Pure(v) if isMainLevel =>
+                  translateStep(view.next(r), isMainLevel)
 
-      case failed: FreeC.Result.Fail[Algebra[F, O, ?], Unit] =>
-        Algebra.raiseError(failed.error)
+                case r @ Result.Pure(v) if !isMainLevel =>
+                  // Cast is safe here, as at this point the evaluation of this Step will end
+                  // and the remainder of the free will be passed as a result in Bind. As such
+                  // next Step will have this to evaluate, and will try to translate again.
+                  view.next(r).asInstanceOf[FreeC[Algebra[G, X, ?], Unit]]
 
-      case interrupted: FreeC.Result.Interrupted[Algebra[F, O, ?], _, Unit] =>
-        FreeC.interrupted(interrupted.context, interrupted.deferredError)
+                case r @ Result.Fail(err) => translateStep(view.next(r), isMainLevel)
 
-      case view: ViewL.View[Algebra[F, O, ?], y, Unit] =>
-        view.step match {
-          case output: Algebra.Output[F, O] =>
-            Algebra.output[G, O](output.values).transformWith { r =>
-              translate0(fK, view.next(r), concurrent)
-            }
+                case r @ Result.Interrupted(_, _) => translateStep(view.next(r), isMainLevel)
+              }
 
-          case step: Algebra.Step[F, x, O] =>
-            FreeC
-              .Eval[Algebra[G, O, ?], Option[(Chunk[x], Token, FreeC[Algebra[G, x, ?], Unit])]](
-                Algebra.Step[G, x, O](
-                  stream = translateStep[F, G, x](fK, step.stream, concurrent),
+            case step: Algebra.Step[F, x, X] =>
+              evalCont[Option[(Chunk[x], Token, FreeC[Algebra[G, x, ?], Unit])]] {
+                Algebra.Step[G, x, X](
+                  stream = translateStep[x](step.stream, false),
                   scope = step.scope
-                ))
-              .transformWith { r =>
-                translate0(fK, view.next(r.asInstanceOf[Result[y]]), concurrent)
+                )
               }
 
-          case alg: Algebra.AlgEffect[F, r] =>
-            FreeC
-              .Eval[Algebra[G, O, ?], r](translateAlgEffect(alg, concurrent, fK))
-              .transformWith(r => translate0(fK, view.next(r), concurrent))
+            case alg: Algebra.AlgEffect[F, res] => {
```

Comment:
With the changes in this PR, this is the only use for the `AlgEffect` middle trait. If we unfold the 6 inner cases to the `view.step match{` level, we can remove `AlgEffect`.